File: ports/java/docproc/src/main/kotlin/net/russiandocs/docproc/inference/DeviceResolution.kt
package net.russiandocs.docproc.inference

import net.russiandocs.docproc.pipeline.Device
import java.io.File

/**
 * Deciding which device to actually use, and refusing to find out the hard way.
 *
 * Two rules, both carried from `service/ml/runtime.py` and both re-proved by the Go port:
 *
 * 1. **A listed CUDA provider does not mean a working GPU.** `getAvailableProviders()` reports what the
 *    artifact was BUILT with, so the GPU jar lists CUDA on a machine with no driver at all.
 * 2. **In a container started without `--gpus`, creating a CUDA session SEGFAULTS rather than throwing.**
 *    No try/catch can save the process from that, which is why the device-node probe runs FIRST.
 */
public object DeviceResolution {

    /**
     * Whether a GPU is visible to THIS process.
     *
     * Checks for the device nodes rather than asking the driver, because the question is precisely whether
     * the container was given access — and a library call that answers it is the call that crashes.
     *
     * `/dev/nvidiactl` and `/dev/nvidia0` are the Linux passthrough; `/dev/dxg` is WSL2's. On Windows the
     * probe returns whatever the provider list says, because there is no container boundary to be on the
     * wrong side of and no device node to inspect.
     */
    @JvmStatic
    public fun gpuVisible(): Boolean {
        if (isWindows()) {
            return providerList().any { it.contains("CUDA") }
        }
        // File.exists() is enough here: these are character devices, and unlike some APIs the JVM's
        // File.exists does report them. (In .NET the equivalent needed Path.Exists rather than
        // File.Exists, which answers false for a device node — worth noting because the two look alike.)
        return listOf("/dev/nvidiactl", "/dev/nvidia0", "/dev/dxg").any { File(it).exists() }
    }

    /** The providers ONNX Runtime ADVERTISES. Not what it can actually bind — see the type note. */
    @JvmStatic
    public fun providerList(): List<String> =
        try {
            ai.onnxruntime.OrtEnvironment.getAvailableProviders().map { it.getName() }
        } catch (e: Throwable) {
            // A runtime that cannot even enumerate its providers is a CPU-only situation as far as the
            // caller is concerned; the real failure surfaces when a session is built.
            listOf("CPUExecutionProvider")
        }

    /**
     * Resolves a requested device into an ordered list of attempts.
     *
     * `auto` becomes GPU-then-CPU when a device is visible and CPU alone when it is not. An explicit `gpu`
     * that cannot be honoured falls back to CPU with a loud message rather than crashing — the reference logs
     * and continues, and a service that refuses to start cannot explain why it refused.
     *
     * An explicit `cpu` is honoured EXACTLY and probes nothing. "Give me CPU" is sometimes a correctness
     * requirement rather than a preference: the conformance goldens were generated on CPU.
     */
    @JvmStatic
    public fun resolve(requested: String, log: (String) -> Unit = {}): List<Device> = when (requested) {
        "cpu" -> listOf(Device.CPU)
        "gpu" -> if (gpuVisible()) {
            listOf(Device.GPU, Device.CPU)
        } else {
            log(
                "compute_device=gpu but no GPU is visible to this process — refusing to attempt CUDA, " +
                    "which would TERMINATE the process rather than fail cleanly. Using CPU. " +
                    "In Docker, pass --gpus all.",
            )
            listOf(Device.CPU)
        }
        "auto" -> if (gpuVisible()) listOf(Device.GPU, Device.CPU) else listOf(Device.CPU)
        else -> throw IllegalArgumentException(
            "device must be auto, cpu or gpu, got \"$requested\"")
    }

    private fun isWindows(): Boolean =
        System.getProperty("os.name").orEmpty().startsWith("Windows", ignoreCase = true)
}

File: ports/dotnet/src/RussianDocs.DocumentProcessing/Inference/DeviceResolution.cs
using System.Runtime.InteropServices;

namespace RussianDocs.DocumentProcessing.Inference;

/// <summary>
/// Deciding which device to actually use.
///
/// <para>
/// This is the "GPU if available, else CPU" logic, and the reason it is more than a try/catch is
/// worth stating plainly: **in a container started without <c>--gpus</c>, the CUDA provider
/// SEGFAULTS rather than throwing.** No exception handler can rescue that, so the device must be
/// probed before the provider is ever asked for.
/// </para>
/// </summary>
public static class DeviceResolution
{
    /// <summary>
    /// Whether a GPU is visible to THIS process.
    ///
    /// <para>
    /// On Windows the answer is always yes: the driver is a system component and there is no device
    /// node to look for, so the attempt loop is the only check available. On Linux the device nodes
    /// are the honest signal — and they are what distinguishes "the host has a GPU" from "this
    /// container was given one", which is the case that segfaults.
    /// </para>
    ///
    /// <para>
    /// <c>/dev/dxg</c> is in the list for WSL2, where CUDA works through a paravirtualised device
    /// rather than <c>/dev/nvidia0</c>.
    /// </para>
    /// </summary>
    public static bool GpuVisible()
    {
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
        {
            return true;
        }
        foreach (string node in new[] { "/dev/nvidiactl", "/dev/dxg", "/dev/nvidia0" })
        {
            // Path.Exists rather than File.Exists: these are CHARACTER DEVICES, and File.Exists
            // returns false for them on some runtimes.
            if (Path.Exists(node))
            {
                return true;
            }
        }
        return false;
    }

    /// <summary>
    /// Resolves the requested device to one that works, and says WHY.
    ///
    /// <para>
    /// The reason string is returned rather than logged here, so the caller can put it in its own log
    /// with its own prefix — and so this function stays testable without a logger.
    /// </para>
    /// </summary>
    /// <param name="requested">What the operator asked for.</param>
    /// <param name="probe">
    /// Builds something on a device and throws if it cannot. Passed in rather than called directly
    /// because "usable" means "the twelve real sessions construct", which only the caller knows how
    /// to attempt.
    /// </param>
    public static (Device Device, string Reason) Resolve(Device requested, Action<Device> probe)
    {
        // **A CPU request is honoured exactly, with no probing.** Sometimes "give me CPU" is a
        // correctness requirement rather than a preference — the conformance goldens were generated
        // on CPU — so silently upgrading it would be wrong.
        if (requested != Device.Gpu)
        {
            probe(Device.Cpu);
            return (Device.Cpu, "cpu requested");
        }

        if (!GpuVisible())
        {
            probe(Device.Cpu);
            return (Device.Cpu,
                "gpu requested but no device is visible to this process; CUDA was NOT attempted, " +
                "because without a device the provider can terminate the process instead of " +
                "returning an error");
        }

        try
        {
            probe(Device.Gpu);
            return (Device.Gpu, "gpu requested and available");
        }
        catch (Exception gpuError)
        {
            // A listed provider is not a working GPU: CUDAExecutionProvider appears whenever the GPU
            // package is installed, with no driver and no cuDNN present. Falling back is the point of
            // the loop.
            try
            {
                probe(Device.Cpu);
            }
            catch (Exception cpuError)
            {
                throw new InvalidOperationException(
                    $"inference: neither device usable; cpu: {cpuError.Message} " +
                    $"(gpu: {gpuError.Message})", cpuError);
            }
            return (Device.Cpu, $"gpu requested but unusable, fell back to cpu: {gpuError.Message}");
        }
    }
}
